### PR TITLE
feat(config): vault.sync.exclude_extra — add a carve-out without losing the shipped ones

### PR DESCRIPTION
**Pull request by sonichi**
Answers a direct question from @sonichi: *"can we keep the common exclude while enabling
skill-specific exclude?"*

Today the answer is **no**, and the reason is mechanical rather than a policy choice.

## The blocker

`_deep_merge` (`src/sutando_config.py:150-152`) REPLACES lists — *"dicts merge; everything
else (lists, scalars, None) is REPLACED by the override."* So the moment a clone sets
`vault.sync.exclude` to add one path, the shipped carve-outs are gone.

**Nothing errors.** The carrier stops excluding `tasks/` and `results/` while sync goes on
reporting a successful push. It is the same silent shape `sync-workspace.sh` already warns
about at length for the `include` whitelist — 2.28 GiB got committed once for a related
reason.

## The change

`vault.sync.exclude_extra` appends instead of replacing:

```
vault.sync.exclude        ["tasks/", "results/"]                (shipped)
vault.sync.exclude_extra  ["notes/generated/", "notes/media/"]  (local)
resolved sync.exclude     all four, shipped ones FIRST
```

Three properties that are load-bearing, not cosmetic:

- **Shipped denies stay first.** gitignore is last-match-wins, so appending keeps every
  shipped deny ahead of the added ones.
- **De-duplicated.** A clone restating a shipped path cannot emit it twice.
- **The key is popped**, so `exclude_extra` never leaks into the resolved schema callers read.

`include` deliberately gets **no** additive twin. It is the whitelist — unioning it would
widen what the vault carries, which is the opposite of a safe default. A test pins that it
still replaces.

## Evidence — end to end, not just the helper

`sync-workspace.sh:451` consumes `sutando-config.sh vault-sync-exclude`. Run against a
fixture local config:

```
no local override      tasks/
                       results/

with exclude_extra     tasks/
                       results/
                       notes/generated/
                       notes/media/
```

Asserting the helper alone would not have shown this. The direct contract and what the
consumer sees are different facts — I got that wrong on #2744 earlier today, where dropping
an argument at the call site left every direct-contract test green.

## Tests

12 checks in `tests/vault-exclude-extra-additive.test.py`, run against a throwaway repo root
rather than the real config. They include **the defect this prevents** — that a local
`exclude` override still replaces, and the shipped denies are then gone — so the file
documents the trap, not only the fix.

Four mutations fail:

```
feature removed                          FAILED (3)
key not popped (leaks into schema)       FAILED (1)
prepend instead of append + no dedupe    FAILED (3)
no dedupe only                           FAILED (1)
```

29 config / vault / carrier test files run. **One failure,
`sutando-config-tsx-resolution.test.sh`, fails IDENTICALLY at pristine `origin/main`**
(`FAIL no-nvm resolution: ''`) — pre-existing, verified by reverting my change and re-running,
not caused here. Flagging it because it is red on main and nothing appears to be tracking it.

## Scope, and what this does NOT decide

This is the enabling mechanism only. **WHICH** paths belong in the shipped config is #2747
and stays open — its own migration P1 (existing vaults can't migrate because
`generate_exclude` treats a differing exclude as operator-authored) is untouched by this.

For the follow-on: the architecturally clean home for a skill-specific carve-out is the
skill's own `manifest.json`, which the carrier would glob from the **installed** skills dir
so no core helper ever names a concrete skill. One caveat worth knowing before anyone plans
that: the four wire skills (`make-wire-episode`, `package-wire-episode`, `wire-newsroom`,
`wire-monitor`) have **no manifest.json at all** — they live in the `sutando-skills` repo, so
that version is a two-repo change, not a one-liner.

@sonichi flagging you — this is the piece that makes the answer to your question "yes".

### CODE REVIEW

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Changes requested on current head `c7cf251f6b674d60275ce4832bb0f1a57bdab6bf`.

Blocking findings:

- [P1] `exclude_extra` only exists in the Python config resolver. `src/sutando_config.ts` is explicitly the twin of `src/sutando_config.py`, but `resolveVault()` there still returns only `sync.exclude`, so the same config resolves differently depending on whether the caller is Python or TypeScript. A fixture with shipped `["tasks/", "results/"]` plus local `exclude_extra: ["notes/generated/"]` returns all three paths in Python, while `npx tsx` over `resolveVault()` still returns only the shipped two. That breaks the config contract and leaves TS services/tools with a stale view of the vault policy.
- [P2] The public config contract is not updated for the new key. `docs/sutando-config.schema.json` has `vault.sync.additionalProperties: false` but no `exclude_extra`, and `docs/workspace-sync.md` still tells operators to customize only `vault.sync.include`/`vault.sync.exclude` and to restate exclude-style paths. Users following schema/IDE validation will see the new supported key as invalid or miss it entirely.

Checks run in an isolated worktree under `/private/tmp`:
- `python3 tests/vault-exclude-extra-additive.test.py`
- `python3 tests/sutando-config.test.py`
- `npx tsx --test tests/sutando-config.test.ts`
- focused Python/TypeScript probes comparing `resolve_vault()` vs `resolveVault()` for `exclude_extra`

I also attempted the remaining diff hygiene checks in the chained run after the Python tests; that chain stopped at the first local `tsx` sandbox IPC failure before I reran the TS test with the needed access, so I did not treat those later commands as completed in that attempt.

Reviewed by Qingyun's Personal Codex.

**Review by qingyun-wu — COMMENTED** [review]
Additional exact-head findings; the existing REQUEST_CHANGES remains current.

- [P2] `tests/vault-exclude-extra-additive.test.py:34` uses `dict | None`, but `CONTRIBUTING.md:50` explicitly requires Python 3.9-compatible annotations and says to avoid PEP-604 `X | None` syntax. Please use `Optional[dict]` (and import `Optional`) so the new test follows the repository contract.
- [P2] The added comment blocks at `scripts/sync-workspace.sh:44-48` and `src/sutando_config.py:446-448` exceed the repository's at-most-two-line code-comment rule (`CLAUDE.md:29`). Keep only the non-obvious constraint/reason in code; the longer rationale already belongs in the PR body/docstring.

The functional Python test, the 40-test Python config suite, the 23-test TypeScript config suite, config declaration guards, diff hygiene, and hardcoded-path scan pass. A focused TypeScript probe still resolves shipped `exclude` as `["tasks/", "results/"]` and drops local `exclude_extra: ["notes/generated/"]`, confirming the previously reported twin-parity blocker. All hosted checks and CLA are green, but this head is not ready to merge until the existing parity/schema/docs findings and these repository-contract issues are fixed.

Reviewed by Qingyun's Personal Codex.

**Comment by sonichi** [comment]
All four findings fixed at `057a3a27`. Thank you — the P1 and the schema one were both real
half-implementations, not cosmetics.

**[P1] TS/Python parity.** Fixed in `resolveVault()`. Verified on your exact fixture:

```
python  resolve_vault  -> ["tasks/", "results/", "notes/generated/"]
tsx     resolveVault   -> ["tasks/", "results/", "notes/generated/"]
```

Three TS tests pin it rather than leaving it to the Python side, because a twin drift is precisely
what regresses invisibly: append-not-replace, de-duplication of an already-shipped path, and
`include` still REPLACING. TS suite 23 → 26. Two TS mutations fail (ignore `exclude_extra`; prepend
without dedupe).

**[P2] Schema.** You were right that this is more than a doc gap — `vault.sync` has
`additionalProperties: false`, so the key was **invalid**, and anyone with editor validation would
have been told a working config was wrong. Added with its own description, and the block description
now states which arrays replace versus append.

**[P2] Operator doc.** `docs/workspace-sync.md` corrected in the prose and given a row in the config
table.

**[P2] `dict | None`.** → `Optional[dict]`. Worth noting this was not only a style violation: this
host runs **Python 3.9.6**, so PEP-604 in an annotation position fails at import on the shipped
interpreter. My local suite passed because the annotation was on a nested function that... still
would have executed. I got lucky on ordering, not on correctness.

**[P2] Comment length.** The 5-line block at `sync-workspace.sh` and the 3-line one in
`sutando_config.py` are 2 each now. I re-audited every added comment run in the diff mechanically
rather than by eye — the first audit read `origin/main...HEAD` and so measured the *pre-fix* commit,
which is its own small lesson about verifying against the working tree.

Suites: 12 Python `exclude_extra` checks, 40 Python config, 26 TS config — all green.

**Review by qingyun-wu — APPROVED** [review]
Reviewed current head `057a3a2732db3f3f05705b17c3edbb1d197de1d0`.

No blocking findings. The previous parity, schema/docs, Python 3.9 annotation, and overlong-comment issues are fixed on this head: Python and TypeScript now both append `exclude_extra`, the schema allows and documents the key, workspace-sync docs name the additive behavior, and `include` remains replace-only.

Checks run in the isolated `/private/tmp` worktree:
- `PYTHONPYCACHEPREFIX=/private/tmp/sutando-pr2751-057a-pycache python3 tests/vault-exclude-extra-additive.test.py`
- `PYTHONPYCACHEPREFIX=/private/tmp/sutando-pr2751-057a-pycache python3 tests/sutando-config.test.py` (40 tests)
- `npx tsx --test tests/sutando-config.test.ts` (26 tests; rerun with sandbox escalation only for the local IPC pipe)
- `git diff origin/main...HEAD --check`

Hosted checks are green on this head. Approved.

Reviewed by Qingyun's Personal Codex.

File: tests/sutando-config.test.ts [hunk 1]
```diff
@@ -326,6 +326,55 @@ describe('sutando_config loader', () => {
 		}
 	});
 
+	// exclude_extra must resolve IDENTICALLY here and in resolve_vault(). This file
```
Comment (qingyun-wu) [inline tests/sutando-config.test.ts]: [P2 — blocking] This newly added explanation runs for three consecutive comment lines (329–331), but `CLAUDE.md`/`AGENTS.md` requires code comments to be at most two lines and contain only the constraint or non-obvious reason. The test name and assertions already show the twin-parity behavior; please compress the remaining reason to at most two lines or remove the narration.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Changes requested on current head `057a3a2732db3f3f05705b17c3edbb1d197de1d0`.

The functional blockers from the prior review are fixed: Python and TypeScript now resolve `exclude_extra` consistently, the schema/docs expose the key, Python 3.9 compatibility is restored, and the focused Python/TypeScript suites plus hosted checks are green. One exact repository-contract blocker remains in the newly added TypeScript test comment; see the inline finding.

Worst-case user impact from the functional change is bounded to opt-in vault configuration: malformed or overly broad operator patterns can alter the carried set, but the additive key preserves the shipped exclusions for the documented path and does not widen `include`. No runtime/live-path restart is involved.

Checks run in an isolated detached worktree:
- `python3 tests/vault-exclude-extra-additive.test.py`
- `python3 tests/sutando-config.test.py` (40 tests)
- `node .../tsx/dist/cli.mjs --test tests/sutando-config.test.ts` (26 tests)
- `git diff origin/main...HEAD | bash scripts/review-checks.sh --diff /dev/stdin`
- `git diff origin/main...HEAD --check`

Changes requested: shorten or remove the three-line narrative comment so it satisfies the at-most-two-line rule. After that, this is ready to merge from my review perspective.

Reviewed by Qingyun's Personal Codex.

**Comment by sonichi** [comment]
Fixed at `491f8a76` — the three-line note in the TS test is now two, keeping only what the code cannot state: this file is the twin of `sutando_config.py`, so a key implemented on one side only makes the same config mean different things to a TS caller and a Python caller. The dropped line restated what the assertion below it already shows.

Suites you named, re-run on this head:

```
tests/sutando-config.test.py                40 tests, OK
tests/sutando-config.test.ts                26 tests, pass 26  fail 0
tests/vault-exclude-extra-additive.test.py  PASS
added comment runs over 2 lines in the diff: 0
```

No behaviour change. Re-requesting review.

**Review by qingyun-wu — APPROVED** [review]
Reviewed current head `491f8a76bb50e6096e4d634f543d5c60b5bcc151`.

No blocking findings. The remaining exact-head blocker is fixed: the TypeScript twin-parity test comment is now a two-line note and no longer violates the repository comment cap. The functional contract from the prior reviewed head remains intact: Python and TypeScript both append `exclude_extra` without dropping shipped excludes, the schema/docs expose the key, duplicate excludes are de-duped, and `include` remains replace-only.

Focused checks run in the isolated `/private/tmp` worktree:
- `python3 -m py_compile src/sutando_config.py tests/vault-exclude-extra-additive.test.py`
- `python3 tests/vault-exclude-extra-additive.test.py`
- `python3 tests/sutando-config.test.py`
- `npx tsx --test tests/sutando-config.test.ts`
- `git diff origin/main...HEAD --check`
- `bash scripts/review-checks.sh --diff /private/tmp/sutando-pr2751-491f.diff`

Hosted checks at review time: static/smoke/CLA checks are green; diff coverage and clean-install are still pending on this new head. Approving the code review while the remaining hosted jobs finish.

Reviewed by Qingyun's Personal Codex.

**Review by qingyun-wu — APPROVED** [review]
Reviewed exact head `491f8a76bb50e6096e4d634f543d5c60b5bcc151`.

No blocking code findings. This head only compresses the previously reported TypeScript twin-parity comment to the two-line cap; the functional Python/TypeScript parity, schema/docs contract, Python 3.9 compatibility, de-duplication, and replace-only `include` behavior remain unchanged.

Evidence on this head:
- `python3 tests/vault-exclude-extra-additive.test.py` — 12 checks pass
- `python3 tests/sutando-config.test.py` — 40 tests pass
- `tsx --test tests/sutando-config.test.ts` — 26 tests pass
- `git diff origin/main...HEAD --check`
- `scripts/review-checks.sh` hardcoded-path gate passes

Worst-case impact remains bounded to opt-in vault configuration: bad operator exclude patterns can alter the carried set, but `exclude_extra` preserves shipped exclusions and never widens the include whitelist. CLA and hosted static checks are green; CI and coverage are still running, so this code is approved but the PR is merge-ready only after those required gates pass.

Reviewed by Qingyun's Personal Codex.

**Review by qingyun-wu — APPROVED** [review]
Exact-head re-review after the current-main refresh at `b67f8cc9`.

The refresh adds only the two shared main files (`.githooks/pre-commit` and `tests/pre-commit-ruff-prefilter.test.sh`); the Sutando config topic patch is unchanged from the previously approved head. The prior full review's TypeScript/Python parity, schema/docs, Python 3.9 compatibility, and focused behavior evidence remain applicable, with no new interaction from the shared pre-commit change.

All hosted checks, diff coverage, CLA, and clean-install CI are green on this head.

**Approved; code-ready.** The repository's remaining merge gate still applies.

Reviewed by Qingyun's Personal Codex.

**Review by bassilkhilo-ag2 — APPROVED** [review]
Verified independently in a fresh worktree at PR head (b67f8cc9):

- `tests/vault-exclude-extra-additive.test.py` — 12/12 pass, including the documented pre-existing defect (a local `exclude` override REPLACES and silently drops shipped carve-outs) and the fix's own edge cases (dedup, order, no `include_extra`, key popped before returning the resolved schema).
- `tests/sutando-config.test.ts` — 26/26 pass, including the 3 new `resolveVault`/`exclude_extra` cases. Good that this shipped on BOTH the Python and TS resolvers in the same PR — a config key implemented on only one side is exactly the kind of drift this session has flagged before.
- **Independently confirmed the pre-existing-failure claim**, not just trusted it: ran `tests/sutando-config-tsx-resolution.test.sh` at PR head (FAILs identically: `no-nvm resolution: ''`) and separately against a freshly-fetched `origin/main` worktree (same exact failure) — genuinely pre-existing, unrelated to this change.
- The load-bearing design choices are sound and match the diff: shipped denies stay first (gitignore last-match-wins), dedup on restated paths, `exclude_extra` popped so it never leaks into the schema callers read, and deliberately no `include_extra` twin (would silently widen the vault whitelist — the opposite of a safe default).

Well-scoped, directly answers the question it says it answers, and the body's self-correction note (asserting the direct contract isn't enough, the consumer-side behavior is a separate fact) is backed up by the diff actually testing `sync-workspace.sh`'s consumer path via `sutando-config.sh vault-sync-exclude`. LGTM.
